Source/WebCore/rendering/MarkedText.cpp

```diff
@@ -113,19 +113,42 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
         auto& parentRenderer = *renderer.parent();
         auto& parentStyle = parentRenderer.style();
         if (auto highlightRegister = renderer.document().highlightRegisterIfExists()) {
-            for (auto& highlight : highlightRegister->map()) {
-                auto renderStyle = parentRenderer.getUncachedPseudoStyle({ PseudoId::Highlight, highlight.key }, &parentStyle);
+            for (auto& highlightName : highlightRegister->set()) {
+                auto renderStyle = parentRenderer.getUncachedPseudoStyle({ PseudoId::Highlight, highlightName }, &parentStyle);
                 if (!renderStyle)
                     continue;
                 if (renderStyle->textDecorationsInEffect().isEmpty() && phase == PaintPhase::Decoration)
                     continue;
-                for (auto& rangeData : highlight.value->rangesData()) {
+                for (auto& rangeData : highlightRegister->map().get(highlightName)->rangesData()) {
                     if (!highlightData.setRenderRange(rangeData))
                         continue;
 
                     auto [highlightStart, highlightEnd] = highlightData.rangeForTextBox(renderer, selectableRange);
-                    if (highlightStart < highlightEnd)
-                        markedTexts.append({ highlightStart, highlightEnd, MarkedText::Highlight, nullptr, highlight.key });
+                    if (highlightStart < highlightEnd) {
+                        // if we can just append it to the end, do that instead
+                        if (markedTexts.size() == 0
+                            || markedTexts[markedTexts.size()-1].priority <= highlightRegister->map().get(highlightName)->priority()) {
+                            markedTexts.append({ highlightStart, highlightEnd, MarkedText::Highlight, nullptr, highlightName, highlightRegister->map().get(highlightName)->priority() });
+                        } else {
+                            // find the first correct place to insert highlight
+                            bool inserted = false;
+                            for (int8_t index = markedTexts.size()-1; index >= 0; index--) {
+                                if (markedTexts[index].priority < highlightRegister->map().get(highlightName)->priority()) {
+                                    markedTexts.insert(index, { highlightStart, highlightEnd, MarkedText::Highlight, nullptr, highlightName, highlightRegister->map().get(highlightName)->priority() });
+                                    inserted = true;
+                                    break;
+                                }
+                                if (markedTexts[index].priority == highlightRegister->map().get(highlightName)->priority()) {
```

Comment:
Could probably save `highlightRegister->map().get(highlightName)->priority)(` into a variable and just re-use that variable instead of repeating this call chain